File: app/src/main/java/com/boogie/vibestation/util/PlaylistUtil.java
package com.boogie.vibestation.util;

import android.content.Context;
import android.content.SharedPreferences;
import android.net.Uri;
import android.os.Handler;
import android.os.Looper;
import android.widget.Toast;

import com.boogie.vibestation.models.Playlist;
import com.boogie.vibestation.models.Song;

import org.json.JSONArray;
import org.json.JSONObject;

import java.io.BufferedReader;
import java.io.InputStream;
import java.io.InputStreamReader;
import java.io.OutputStream;
import java.util.ArrayList;
import java.util.List;
import java.util.Locale;
import java.util.Map;

/**
 * Handles serialization, persistence, backup export, and restoration for playlists.
 */
public final class PlaylistUtil {

    private PlaylistUtil() {
        // Prevent instantiation
    }

    /**
     * Serializes playlist collection structures to JSON array and saves to SharedPreferences.
     *
     * @param prefs Target SharedPreferences instance.
     * @param playlists Playlists to persist.
     */
    public static void savePlaylists(SharedPreferences prefs, List<Playlist> playlists) {
        try {
            JSONArray playlistsJsonArray = new JSONArray();
            for (Playlist playlist : playlists) {
                JSONObject playlistJsonObject = new JSONObject();
                playlistJsonObject.put("name", playlist.name);
                playlistJsonObject.put("imageUri", playlist.imageUri != null ? playlist.imageUri : "");
                playlistJsonObject.put("description", playlist.description != null ? playlist.description : "");
                playlistJsonObject.put("isFire", playlist.isFire);

                JSONArray songsJsonArray = new JSONArray();
                for (Song song : playlist.songs) {
                    JSONObject songJsonObject = new JSONObject();
                    songJsonObject.put("id", song.id);
                    songJsonObject.put("t", song.title);
                    songJsonObject.put("a", song.artist);
                    songsJsonArray.put(songJsonObject);
                }
                playlistJsonObject.put("songData", songsJsonArray);
                playlistsJsonArray.put(playlistJsonObject);
            }
            prefs.edit().putString("playlists", playlistsJsonArray.toString()).apply();
        } catch (Exception ignored) {}
    }

    /**
     * Parses stored playlist JSON configuration from preferences and associates tracks using lookup maps.
     *
     * @param prefs Source SharedPreferences instance.
     * @param songIdMap Lookup map keyed by track ID.
     * @param songNameMap Lookup map keyed by normalized title and artist.
     * @return List of reconstructed Playlist instances.
     */
    public static ArrayList<Playlist> parsePlaylists(SharedPreferences prefs, Map<String, Song> songIdMap, Map<String, Song> songNameMap) {
        ArrayList<Playlist> tempPlaylists = new ArrayList<>();
        try {
            JSONArray playlistsJsonArray = new JSONArray(prefs.getString("playlists", "[]"));
            for (int i = 0; i < playlistsJsonArray.length(); i++) {
                JSONObject playlistJsonObject = playlistsJsonArray.getJSONObject(i);
                Playlist playlist = new Playlist(
                        playlistJsonObject.getString("name"),
                        playlistJsonObject.optString("imageUri", null)
                );
                playlist.description = playlistJsonObject.optString("description", "");
                playlist.isFire = playlistJsonObject.optBoolean("isFire", false);
                JSONArray songsJsonArray = playlistJsonObject.getJSONArray("songData");
                for (int j = 0; j < songsJsonArray.length(); j++) {
                    JSONObject songJsonObject = songsJsonArray.getJSONObject(j);
                    String songId = songJsonObject.getString("id");
                    String songTitle = songJsonObject.getString("t");
                    String songArtist = songJsonObject.getString("a");

                    Song matchedSong = songIdMap.get(songId);
                    if (matchedSong == null) {
                        matchedSong = songNameMap.get((songTitle + "_" + songArtist).toLowerCase(Locale.getDefault()));
                    }
                    if (matchedSong != null) {
                        playlist.songs.add(matchedSong);
                    }
                }
                tempPlaylists.add(playlist);
            }
        } catch (Exception ignored) {}
        return tempPlaylists;
    }

    /**
     * Serializes playlist collection and embedded base64 artwork into a backup file on a background thread.
     *
     * @param context Application context for Toast and content resolution.
     * @param prefs Source SharedPreferences instance.
     * @param documentUri Target file URI for writing backup data.
     */
    public static void exportBackup(Context context, SharedPreferences prefs, Uri documentUri) {
        if (documentUri == null) return;
        new Thread(() -> {
            try (OutputStream outputStream = context.getContentResolver().openOutputStream(documentUri)) {
                JSONArray playlistsJsonArray = new JSONArray(prefs.getString("playlists", "[]"));
                for (int i = 0; i < playlistsJsonArray.length(); i++) {
                    JSONObject playlistJsonObject = playlistsJsonArray.getJSONObject(i);
                    String imageUri = playlistJsonObject.optString("imageUri", "");
                    if (!imageUri.isEmpty()) {
                        playlistJsonObject.put("b64", ArtUtil.getBase64Image(context.getContentResolver(), Uri.parse(imageUri)));
                    }
                }
                if (outputStream != null) {
                    outputStream.write(playlistsJsonArray.toString().getBytes());
                }
                Handler handler = new Handler(Looper.getMainLooper());
                handler.post(() -> Toast.makeText(context, "Export Ready!", Toast.LENGTH_SHORT).show());
            } catch (Exception ignored) {}
        }).start();
    }

    /**
     * Parses and restores playlists JSON backup data from a user-selected document.
     *
     * @param context Application context for content resolution.
     * @param prefs Target SharedPreferences instance.
     * @param documentUri Selected backup document URI.
     * @param onRestoreComplete Callback executed on UI thread upon successful restoration.
     */
    public static void restoreBackup(Context context, SharedPreferences prefs, Uri documentUri, Runnable onRestoreComplete) {
        if (documentUri == null) return;
        new Thread(() -> {
            try (InputStream inputStream = context.getContentResolver().openInputStream(documentUri);
                 BufferedReader bufferedReader = new BufferedReader(new InputStreamReader(inputStream))) {
                StringBuilder stringBuilder = new StringBuilder();
                String line;
                while ((line = bufferedReader.readLine()) != null) {
                    stringBuilder.append(line);
                }
                prefs.edit().putString("playlists", stringBuilder.toString()).apply();
                Handler handler = new Handler(Looper.getMainLooper());
                handler.post(() -> {
                    if (onRestoreComplete != null) {
                        onRestoreComplete.run();
                    }
                    Toast.makeText(context, "Restore Successful!", Toast.LENGTH_SHORT).show();
                });
            } catch (Exception ignored) {}
        }).start();
    }
}

File: app/src/main/kotlin/com/boogie/vibestation/util/PlaylistUtil.kt
package com.boogie.vibestation.util

import android.content.Context
import android.content.SharedPreferences
import android.net.Uri
import android.os.Handler
import android.os.Looper
import android.widget.Toast
import com.boogie.vibestation.models.Playlist
import com.boogie.vibestation.models.Song
import org.json.JSONArray
import org.json.JSONObject
import java.io.IOException
import java.util.Locale
import java.util.concurrent.Executors

/**
 * Handles serialization, persistence, backup export, and restoration for playlists.
 */
object PlaylistUtil {

    private const val PLAYLISTS_KEY = "playlists"

    private val backupExecutor = Executors.newSingleThreadExecutor()
    private val mainHandler by lazy { Handler(Looper.getMainLooper()) }

    /**
     * Serializes playlist collection structures to JSON array and saves to SharedPreferences.
     *
     * @param prefs Target SharedPreferences instance.
     * @param playlists Playlists to persist.
     */
    fun savePlaylists(prefs: SharedPreferences, playlists: List<Playlist>) {
        try {
            val playlistsJsonArray = JSONArray()
            for (playlist in playlists) {
                val songsJsonArray = JSONArray()
                for (song in playlist.songs) {
                    songsJsonArray.put(
                        JSONObject()
                            .put("id", song.id)
                            .put("t", song.title)
                            .put("a", song.artist)
                    )
                }
                playlistsJsonArray.put(
                    JSONObject()
                        .put("name", playlist.name)
                        .put("imageUri", playlist.imageUri ?: "")
                        .put("description", playlist.description)
                        .put("isFire", playlist.isFire)
                        .put("songData", songsJsonArray)
                )
            }
            prefs.edit().putString(PLAYLISTS_KEY, playlistsJsonArray.toString()).apply()
        } catch (ignored: Exception) {
        }
    }

    /**
     * Parses stored playlist JSON configuration from preferences and associates tracks using lookup maps.
     *
     * @param prefs Source SharedPreferences instance.
     * @param songIdMap Lookup map keyed by track ID.
     * @param songNameMap Lookup map keyed by normalized title and artist.
     * @return List of reconstructed Playlist instances.
     */
    fun parsePlaylists(
        prefs: SharedPreferences,
        songIdMap: Map<String, Song>,
        songNameMap: Map<String, Song>
    ): ArrayList<Playlist> {
        val playlists = ArrayList<Playlist>()
        try {
            val playlistsJsonArray = JSONArray(prefs.getString(PLAYLISTS_KEY, "[]"))
            for (i in 0 until playlistsJsonArray.length()) {
                val playlistJson = playlistsJsonArray.getJSONObject(i)
                val playlist = Playlist(
                    playlistJson.getString("name"),
                    playlistJson.optString("imageUri", "").takeUnless { it.isBlank() }
                )
                playlist.description = playlistJson.optString("description", "")
                playlist.isFire = playlistJson.optBoolean("isFire", false)

                val songsJsonArray = playlistJson.getJSONArray("songData")
                for (j in 0 until songsJsonArray.length()) {
                    val songJson = songsJsonArray.getJSONObject(j)
                    val nameKey = "${songJson.getString("t")}_${songJson.getString("a")}".lowercase(Locale.getDefault())
                    val matchedSong = songIdMap[songJson.getString("id")] ?: songNameMap[nameKey]
                    if (matchedSong != null) {
                        playlist.songs.add(matchedSong)
                    }
                }
                playlists.add(playlist)
            }
        } catch (ignored: Exception) {
        }
        return playlists
    }

    /**
     * Serializes playlist collection into a backup file on a background thread.
     *
     * @param context Application context for Toast.
     * @param prefs Source SharedPreferences instance.
     * @param documentUri Target file URI for writing backup data.
     */
    fun exportBackup(context: Context, prefs: SharedPreferences, documentUri: Uri?) {
        if (documentUri == null) return
        backupExecutor.execute {
            try {
                context.contentResolver.openOutputStream(documentUri).use { outputStream ->
                    val playlistsJsonArray = JSONArray(prefs.getString(PLAYLISTS_KEY, "[]"))
                    outputStream?.write(playlistsJsonArray.toString().toByteArray(Charsets.UTF_8))
                    toastOnMain(context, "Export Ready!")
                }
            } catch (ignored: Exception) {
            }
        }
    }

    /**
     * Parses and restores playlists JSON backup data from a user-selected document.
     *
     * @param context Application context for content resolution.
     * @param prefs Target SharedPreferences instance.
     * @param documentUri Selected backup document URI.
     * @param onRestoreComplete Callback executed on UI thread upon successful restoration.
     */
    fun restoreBackup(context: Context, prefs: SharedPreferences, documentUri: Uri?, onRestoreComplete: () -> Unit) {
        if (documentUri == null) return
        backupExecutor.execute {
            try {
                val inputStream = context.contentResolver.openInputStream(documentUri)
                    ?: throw IOException("Unable to open backup document")
                val backupContent = inputStream.bufferedReader(Charsets.UTF_8).use { it.readLines().joinToString("") }

                // Validate that the restored file contains a valid JSON array and strip redundant b64 payloads
                val validatedArray = JSONArray(backupContent)
                for (i in 0 until validatedArray.length()) {
                    validatedArray.optJSONObject(i)?.remove("b64")
                }
                prefs.edit().putString(PLAYLISTS_KEY, validatedArray.toString()).apply()
                mainHandler.post {
                    onRestoreComplete()
                    Toast.makeText(context, "Restore Successful!", Toast.LENGTH_SHORT).show()
                }
            } catch (e: Exception) {
                toastOnMain(context, "Failed to restore backup: Invalid file")
            }
        }
    }

    /** Shows a short toast from any thread. */
    private fun toastOnMain(context: Context, message: String) {
        mainHandler.post { Toast.makeText(context, message, Toast.LENGTH_SHORT).show() }
    }
}
